src/java/org/apache/cassandra/db/compaction/CompactionManager.java

```diff
@@ -1219,45 +1267,57 @@ public void forceUserDefinedCleanup(String dataFiles)
             ColumnFamilyStore cfs = Keyspace.open(desc.ksname).getColumnFamilyStore(desc.cfname);
             desc = cfs.getDirectories().find(new File(filename.trim()).name());
             if (desc != null)
-                descriptors.put(cfs, desc);
+                descriptors.computeIfAbsent(cfs, c -> new ArrayList<>()).add(desc);
         }
 
+        descriptors.keySet().forEach(cfs -> info.markKeyspace(cfs.getKeyspaceName()));
         if (!StorageService.instance.isJoined())
         {
             logger.error("Cleanup cannot run before a node has joined the ring");
+            info.result = AllSSTableOpStatus.ABORTED;
             return;
         }
 
-        for (Map.Entry<ColumnFamilyStore,Descriptor> entry : descriptors.entrySet())
+        for (Map.Entry<ColumnFamilyStore, List<Descriptor>> entry : descriptors.entrySet())
         {
             ColumnFamilyStore cfs = entry.getKey();
             Keyspace keyspace = cfs.keyspace;
             final RangesAtEndpoint replicas = StorageService.instance.getLocalReplicas(keyspace.getName());
             final Set<Range<Token>> allRanges = replicas.ranges();
             final Set<Range<Token>> transientRanges = replicas.onlyTransient().ranges();
             boolean hasIndexes = cfs.indexManager.hasIndexes();
-            SSTableReader sstable = lookupSSTable(cfs, entry.getValue());
+            List<SSTableReader> sstables = entry.getValue().stream().
```

Comment:
would you mind to rewrite this to for loops?

https://lists.apache.org/thread/65glsjzkmpktzmns6j9wvr4nczvskx36

Reply:
Fixed. However, I don't think this is a hot path :-)